### ISSUE TITLE
In KDE Plasma, keyring uses gnome-keyring by default

### ISSUE THREAD

**Issue by mczp** [jaraco/keyring#273]
In a relatively fresh install of Fedora 25 with KDE Plasma as the DE, seems keyring prefers to use gnome-keyring (libgnome-keyring) by default.
Getting rid of gnome-keyring doesn't solve the problem, while erasing libgnome-keyring is not feasible as it is a dependency of many packages.
Any chance to add some logic to detect the desktop environment and use kwallet if running under plasma?

**Comment by mitya57** [comment]
Please make sure you have [dbus-python](https://www.freedesktop.org/wiki/Software/DBusBindings/#dbus-pythonbasedondbus-glib) installed. Once you have it, the KWallet backend (and the Secret Storage one) should have higher priority than the libgnome-keyring based one.

**Comment by frispete** [comment]
Unfortunately, Secret Storage still have a higher priority compared to the KWallets. Since Secret Storage is a dependency of python-keyring, it always wins. I'm using ugly code like this
```
        krmap = dict([(kr.name, kr) for kr in keyring.backend.get_all_keyring()])
        kr = None
        for krname in (
            'kwallet DBusKeyringKWallet4',  # kde4
            'kwallet DBusKeyring',          # kde5
            'SecretService Keyring',        # gnome
        ):
            log.debug('check <%s> keyring', krname)
            if krname in krmap:
                kr = krmap[krname]
                break

        if kr is not None:
            self.keyring = kr
        else:
            self.keyring = keyring.get_keyring()
```

As my environments are all KDE, two active keyrings (KDE4 and KDE5) are upsetting my users already...

I haven't had a chance to come up with a better "mergable" approach. Sorry.
The best way would involve to inquire the current DE and adjust priorities accordingly.

Rather OT here, but another issue in this regard is unsolved: keyrings with a graphical password dialogs need to be separated from terminal based ones.

**Comment by mitya57** [comment]
It is intentional that the Secret Service backend has a higher priority. We do not want to use the DE detecting hacks like looking at `$XDG_CURRENT_DESKTOP`, and Secret Service is an XDG standard while KWallet D-Bus API is not.

**Comment by frispete** [comment]
> It is intentional that the Secret Service backend has a higher priority. We do not want to use the DE detecting hacks like looking at $XDG_CURRENT_DESKTOP, and Secret Service is an XDG standard while KWallet D-Bus API is not.

Be assured, that I would have started earlier contributing to this project because of this:
When I first installed keyring, it insisted in using the Secret Storage frontend. I thought: hmm, a GNOME only thing, and abandoned it right away.

Later on, I found multiple hints, that KWallet is supported somehow, but took me reading and adding code in order to use it in my environments, which resulted in the contributed patches adding KDE4 wallet support.

While I understand all sediments about standards et.al., user experience for **all KDE users** is really bad because of this, resulting in reduced propagation of keyring usage. 

*In short: clean code, but less potential users!*

Sure, I don't change anybody's mind in this respect, but wanted to let you know about the consequences of such decisions.

**Comment by mczp** [comment]
I checked a few things, and indeed it makes perfect sense following the XDG spec, I added a manual config option in my app which allows the user to specify forcibly to use any backend.
I hope KDE devs will find time to fix the kwallet stuff the same way the standardized secret storage...

**Comment by jaraco** [comment]
It is a design goal of this project that it detects the best available backend for a given environment. If SecretService is only available for Gnome, and KDE environments are selecting it but unable to use it, that seems inappropriate to me.

What happens to a KDE user when SecretService backend is detected?

**Comment by mitya57** [comment]
> If SecretService is only available for Gnome, and KDE environments are selecting it but unable to use it, that seems inappropriate to me.

SecretService backend is enabled only if `secretstorage.get_all_collections()` returns something, which means the users should be able to use it.

They may not *want* to use it, but that’s what we have configuration files for.

Though I don’t mind if we add a couple of desktop detection lines back somewhere.

**Comment by jaraco** [comment]
> They may not want to use it, but that’s what we have configuration files for.

I agree, and that's fine, if what this project _recommends_ is that KDE users use SecretStorage (which was my understanding). If it's common for a KDE user to want to avoid SecretStorage and use KWallet, then we should provide a documented mechanism for them to readily select KWallet. If it's almost universal for a KDE user to want to avoid SecretStorage and use KWallet, then keyring should give priority to those backends in that environment.

At least, that's my feeling. I don't use Linux environments with keyring, so my opinion is only about the overall expectation for the project.

**Comment by mitya57** [comment]
Should I make a change that bumps KWallet backend priority to 5.1 on Plasma desktop?

**Comment by frispete** [comment]
The average KDE user deals with KWallets already. During the transitioning to KDE Framework 5, the situation is worsen by parts still using KWallet 4 (e.g. stable kdepim packages) and others KWallet5 (dolphin, et.al). Presenting them yet another wallet such as SecretStorage is mostly unwanted, so yes, I vote for bumping KWallets in Plasma environments.

**Comment by frispete** [comment]
@jaraco While at it, what do you think about the terminal vs. graphical password interfaces? Want me to add a property?

**Comment by jaraco** [comment]
@frispete: It's not obvious to me what experience one would desire relating to terminal vs. graphical password interfaces. As a Windows and now Mac user, the distinction was irrelevant. Would you be willing to write up a separate ticket describing the problem and maybe some proposed solutions, so that we can iterate on the concept?

**Comment by takluyver** [comment]
The Secret Service DBus API is supposed to be a standard, which both Gnome Keyring and KDE KWallet implement. The authors of the spec come from those two projects:
https://specifications.freedesktop.org/secret-service/

**Comment by mitya57** [comment]
@takluyver Unfortunately KWallet does not support Secret Service. There was a KSecretService which was intended to be a replacement for KWallet, but its development is stuck now.

**Comment by takluyver** [comment]
That's unfortunate :-(

**Comment by jaraco** [comment]
It feels like there's no progress to be made on this effort, so I'll close for now. Feel free to revive the discussion/effort at any point.

**Comment by piotr-dobrogost** [comment]
> It feels like there's no progress to be made on this effort, so I'll close for now. Feel free to revive the discussion/effort at any point.

What about the following question asked by @mitya57 ?
> Should I make a change that bumps KWallet backend priority to 5.1 on Plasma desktop?

It would implement the following idea:
> If it's almost universal for a KDE user to want to avoid SecretStorage and use KWallet, then keyring should give priority to those backends in that environment.

**Comment by jaraco** [comment]
Was that a question for me? I’m not opposed to tweaking the priorities based on environmental factors. That’s what they’re intended to do.

**Comment by mitya57** [comment]
@jaraco Previously you were opposed to that, e.g. see this changelog entry:
> * Issue #186: Removed preference for keyrings based on
>   ``XDG_CURRENT_DESKTOP`` as these values are too varied
>   to be a reliable indicator of which keyring implementation
>   might be preferable.

I will now try to partially revert the old behavior for Plasma.

**Comment by jaraco** [comment]
Oh man. Sorry to flip-flop. Part of the problem is that I don’t have a good sense for what is right here and only those adversely affected by the current implementation will complain. 

@mitya57 I’ll happily defer to to on whatever approach you think is best.

**Comment by mitya57** [comment]
Unfortunately KDE still did not adapt the Secret Service API (probably because the lack of manpower to do so), they are still using KWallet which is technically inferior and less secure. The last discussion on KWallet vs. Secret Service that I was able to find is [this thread](https://mail.kde.org/pipermail/plasma-devel/2016-July/055641.html) from 2016.

So I think we have to use KWallet on KDE until something changes. I will push my commit from #315 to master later today if you have no objections.
